src/engine/variable.js

```diff
@@ -9,19 +9,45 @@ class Variable {
     /**
      * @param {string} id Id of the variable.
      * @param {string} name Name of the variable.
-     * @param {(string|number)} value Value of the variable.
+     * @param {string} type Type of the variable, one of '' or 'list'
      * @param {boolean} isCloud Whether the variable is stored in the cloud.
      * @constructor
      */
-    constructor (id, name, value, isCloud) {
+    constructor (id, name, type, isCloud) {
         this.id = id || uid();
         this.name = name;
-        this.value = value;
+        this.type = type;
         this.isCloud = isCloud;
+        switch (this.type) {
+        case Variable.SCALAR_TYPE:
+            this.value = 0;
+            break;
+        case Variable.LIST_TYPE:
+            this.value = [];
+            break;
+        default:
+            throw new Error(`Invalid variable type: ${this.type}`);
+        }
     }
 
     toXML () {
-        return `<variable type="" id="${this.id}">${this.name}</variable>`;
+        return `<variable type="${this.type}" id="${this.id}">${this.name}</variable>`;
+    }
+
+    /**
+     * Type representation for scalar variables.
+     * @const {string}
+     */
+    static get SCALAR_TYPE () {
+        return '';
```

Comment:
I think this specific value requires some explanation. "Empty string" is how scratch-blocks denotes the scalar type, basically for backwards compatibility with untyped variables. @rschamp I was wondering if we should have an explicit adapter layer between scratch-blocks variable type strings and the vm variable types, particularly because there are some dangers in using the empty string as a meaningful value (e.g. it is falsey).

Reply:
I agree with your impulse to make our internal representation of the type explicit and distinct from Blockly's. However I think that would add some unnecessary complexity right now. But I think we should keep the adapter idea in mind if it presents itself as a problem down the line.